### PR TITLE
Fix editor performance drop over time due to lingering nested object references

### LINKED ISSUE #11261: Nested hit objects are leaked when parent hit object is edited or deleted
1. Inspect `Playfield.lifetimeEntryMap.Count` while editing a slider shape.
2. See the number is increasing rapidly.

Nested hit objects should be removed when the parent hit object is removed or when the nested hit objects are recreated by `ApplyDefaults`.

It is a visual issue for osu!catch editor because new nested hit objects will be stacked every frame while the parent object is edited, resulting in a massive amount of stacked objects.

### PR DESCRIPTION
**Pull request by peppy**
Supersedes #11262
Closes #11261

PRing on behalf of @ekrctb to continue with review (rebased commits from https://github.com/ppy/osu/compare/master...ekrctb:hitobject-entry-nesting?expand=1)

### CODE REVIEW

**Comment by peppy** [comment]
This branch is working as expected. It doesn't solve all of the editor performance degradation, but does its part towards a resolution. Would appreciate reviews.

File: osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        private void onDefaultsApplied(HitObject hitObject)
+        {
+            if (!childrenMap.Remove(hitObject, out var childEntries))
+                return;
+
+            foreach (var entry in childEntries)
+                Remove(entry);
+
+            childEntries.Clear();
+            childrenMap[hitObject] = childEntries;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:112]: How is this safe? `OnDefaultsApplied` looks to be called after the new nested hitobjects have been created, not before. So if it's clearing `childEntries`, isn't it going to be removing entries of the new nested hitobjects?

**Review by smoogipoo — COMMENTED** [review]
Kind of hard to follow this one. The `Add()`/`Remove()` process needs more inline documentation as there's a lot of relationships.

File: osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs (hunk 1 above)
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:112]: I also don't understand the purpose of this line in general. As far as I understand it, this is what this method does:
- `childrenMap[hitObject] -> childEntries`
- `childEntries.Clear()`
- `childrenMap[hitObject] <- childEntries`
It looks like this operation, the last one in my list above, is a no-op? Am I missing something?

Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:112]: Just so I understand correctly, this seems to only be safe because the order of operations is:
- Hitobject changes.
- `HitObject.ApplyDefaults()
- `HitObject.DefaultsApplied()`
- `HitObjectEntryManager.onDefaultsApplied()`  (clears list)
- Sometime later...
- `DrawableHitObject.Apply()`
- `Playfield.GetPooledDrawableRepresentation()` (for each nested object)
- `HitObjectEntryManager.Add()` (adds to list

If I'm reading things correctly? I still find this pretty dodge and hard to follow, but I don't have any suggestions.

File: osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        /// <summary>
+        /// Stores the list of entries managed by this <see cref="HitObjectEntryManager"/> for each hit object managed by this <see cref="HitObjectEntryManager"/>.
+        /// </summary>
+        private readonly Dictionary<HitObject, List<HitObjectLifetimeEntry>> childrenMap = new Dictionary<HitObject, List<HitObjectLifetimeEntry>>();
+
+        public void Add(HitObjectLifetimeEntry entry, HitObject? parent)
+        {
+            if (parentMap.ContainsKey(entry))
+                throw new InvalidOperationException($@"The {nameof(HitObjectLifetimeEntry)} is already added to this {nameof(HitObjectEntryManager)}.");
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:55]: Any reason for checking `parentMap` here (and in `Remove()`), instead of `entryMap`? Perhaps you could assert that `parentMap` doesn't contain it, but I feel like `entryMap` explicitly saying "each entry" (understanding to mean all entries) makes it more explicit as to what's going on.

File: osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        /// A <c>null</c> is stored for entries of the top-level hit objects.
+        /// </summary>
+        /// <remarks>
+        /// The parent hit object of a pooled hit object may be non-pooled.
+        /// In that case, no corresponding <see cref="HitObjectLifetimeEntry"/> is stored in this <see cref="HitObjectEntryManager"/>.
+        /// </remarks>
+        private readonly Dictionary<HitObjectLifetimeEntry, HitObject?> parentMap = new Dictionary<HitObjectLifetimeEntry, HitObject?>();
+
+        /// <summary>
+        /// Stores the list of entries managed by this <see cref="HitObjectEntryManager"/> for each hit object managed by this <see cref="HitObjectEntryManager"/>.
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:48]:
```suggestion
        /// Stores the list of child entries for each hit object managed by this <see cref="HitObjectEntryManager"/>.
```

File: osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        public event Action<HitObjectLifetimeEntry, HitObject?>? OnEntryRemoved;
+
+        /// <summary>
+        /// Provides the reverse mapping of <see cref="HitObjectLifetimeEntry.HitObject"/> for each entry.
+        /// </summary>
+        private readonly Dictionary<HitObject, HitObjectLifetimeEntry> entryMap = new Dictionary<HitObject, HitObjectLifetimeEntry>();
+
+        /// <summary>
+        /// Stores the parent hit object for entries of the nested hit objects.
+        /// A <c>null</c> is stored for entries of the top-level hit objects.
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:39]: Any reason to omit `null`s altogether?

File: osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:112]: I understand now why this is done the way it is, everything relies on that `.Remove()` call. I'm going to refactor this because it's hard to understand and easy to miss.

Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs:117]: Well, I didn't end up changing this. The idea I had was to not do the whole `.Remove()` stuff altogether, but in hindsight this is probably the more performance way of doing things anyway.

I added what I believe is appropriate commenting along the way, tho.

**Comment by smoogipoo** [comment]
@peppy please check my latest changes and see if you agree.

**Review by smoogipoo — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Comment by peppy** [comment]
Looks good!
